fvm/evm/emulator/emulator.go

```diff
@@ -121,11 +121,16 @@ func (bl *BlockView) DirectCall(call *types.DirectCall) (res *types.Result, err
 	// Call tx tracer
 	if proc.evm.Config.Tracer != nil && proc.evm.Config.Tracer.OnTxStart != nil {
 		proc.evm.Config.Tracer.OnTxStart(proc.evm.GetVMContext(), call.Transaction(), call.From.ToCommon())
-		if proc.evm.Config.Tracer.OnTxEnd != nil {
-			defer func() {
-				proc.evm.Config.Tracer.OnTxEnd(res.Receipt(), err)
-			}()
-		}
+		defer func() {
+			if proc.evm.Config.Tracer.OnTxEnd != nil {
+				receipt := &gethTypes.Receipt{}
+				if res != nil {
+					receipt = res.Receipt()
+				}
```

Comment:
Minor: maybe avoid extra allocation in happy case. Same below

```suggestion
				var receipt *gethTypes.Receipt
				if res != nil {
					receipt = res.Receipt()
				} else {
					receipt = &gethTypes.Receipt{}
				}
```

Reply:
I was not sure the underlying layer could handle the not allocated case.

Reply:
Yeah, allocating a new value in the unhappy path seems reasonable if we don't know if passing down nil is valid. Maybe we can avoid the allocation in the happy path though